File: rust/src/codec.rs
use serde::{Serialize, Deserialize};
use rmpv::Value;

/// Wire messages are encoded as msgpack maps (string keys) so that optional
/// fields can be absent without breaking deserialization. This matches the
/// Python/Go implementations which use msgpack dicts.
pub fn to_vec_map<T: Serialize>(value: &T) -> Result<Vec<u8>, rmp_serde::encode::Error> {
    let mut buf = Vec::new();
    let mut ser = rmp_serde::Serializer::new(&mut buf)
        .with_struct_map();
    value.serialize(&mut ser)?;
    Ok(buf)
}


#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WireMessage {
    pub id: u64,
    #[serde(rename = "type")]
    pub msg_type: String,

    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub func: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub args: Option<Value>,

    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub result: Option<Value>,

    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub error_type: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub message: Option<String>,
}

pub fn pack_request<A: Serialize>(id: u64, func: &str, args: &A) -> Result<Vec<u8>, rmp_serde::encode::Error> {
    #[derive(Serialize)]
    struct RequestRef<'a, A> {
        id: u64,
        #[serde(rename = "type")]
        msg_type: &'static str,
        func: &'a str,
        args: &'a A,
    }

    let req = RequestRef {
        id,
        msg_type: "request",
        func,
        args,
    };
    to_vec_map(&req)
}

pub fn pack_response<R: Serialize>(id: u64, result: &R) -> Result<Vec<u8>, rmp_serde::encode::Error> {
    #[derive(Serialize)]
    struct ResponseRef<'a, R> {
        id: u64,
        #[serde(rename = "type")]
        msg_type: &'static str,
        result: &'a R,
    }
    let resp = ResponseRef {
        id,
        msg_type: "response",
        result,
    };
    to_vec_map(&resp)
}

pub fn pack_error(id: u64, error_type: &str, message: &str) -> Result<Vec<u8>, rmp_serde::encode::Error> {
    #[derive(Serialize)]
    struct ErrorRef<'a> {
        id: u64,
        #[serde(rename = "type")]
        msg_type: &'static str,
        error_type: &'a str,
        message: &'a str,
    }
    let err = ErrorRef {
        id,
        msg_type: "error",
        error_type,
        message,
    };
    to_vec_map(&err)
}

pub fn pack_stream_chunk<R: Serialize>(id: u64, result: &R) -> Result<Vec<u8>, rmp_serde::encode::Error> {
    #[derive(Serialize)]
    struct ChunkRef<'a, R> {
        id: u64,
        #[serde(rename = "type")]
        msg_type: &'static str,
        result: &'a R,
    }
    let chunk = ChunkRef {
        id,
        msg_type: "stream_chunk",
        result,
    };
    to_vec_map(&chunk)
}

pub fn pack_stream_end(id: u64) -> Result<Vec<u8>, rmp_serde::encode::Error> {
    #[derive(Serialize)]
    struct EndRef {
        id: u64,
        #[serde(rename = "type")]
        msg_type: &'static str,
    }
    let end = EndRef {
        id,
        msg_type: "stream_end",
    };
    to_vec_map(&end)
}

/// Decode an inbound wire message. Accepts both map and array formats.
pub fn unpack(payload: &[u8]) -> Result<WireMessage, rmp_serde::decode::Error> {
    rmp_serde::from_slice(payload)
}

File: ts/src/codec.ts
import { encode, decode } from '@msgpack/msgpack';

export interface WireMessage {
  id: number;
  type: string;
  func?: string;
  args?: unknown[];
  stream?: boolean;
  result?: unknown;
  error_type?: string;
  message?: string;
}

/**
 * Pack a request message into a MessagePack buffer.
 */
export function packRequest(id: number, func: string, args: unknown[]): Buffer {
  const msg: WireMessage = { id, type: 'request', func, args };
  return Buffer.from(encode(msg));
}

/**
 * Pack a response message.
 */
export function packResponse(id: number, result: unknown): Buffer {
  const msg: WireMessage = { id, type: 'response', result };
  return Buffer.from(encode(msg));
}

/**
 * Pack an error response message.
 */
export function packError(id: number, errorType: string, message: string): Buffer {
  const msg: WireMessage = { id, type: 'error', error_type: errorType, message };
  return Buffer.from(encode(msg));
}

/**
 * Pack a stream chunk message.
 */
export function packStreamChunk(id: number, result: unknown): Buffer {
  const msg: WireMessage = { id, type: 'stream_chunk', result };
  return Buffer.from(encode(msg));
}

/**
 * Pack a stream_end sentinel.
 */
export function packStreamEnd(id: number): Buffer {
  const msg: WireMessage = { id, type: 'stream_end' };
  return Buffer.from(encode(msg));
}

/**
 * Pack a stream_close sentinel (client-streaming termination).
 */
export function packStreamClose(id: number): Buffer {
  const msg: WireMessage = { id, type: 'stream_close' };
  return Buffer.from(encode(msg));
}

/**
 * Pack a bidirectional-streaming request.
 */
export function packBidiRequest(id: number, func: string, args: unknown[]): Buffer {
  const msg: WireMessage = { id, type: 'request', func, args, stream: true };
  return Buffer.from(encode(msg));
}

/**
 * Decode a raw buffer into a WireMessage.
 */
export function unpack(payload: Buffer): WireMessage {
  const decoded = decode(payload) as Record<string, unknown>;
  return {
    id: decoded.id as number,
    type: decoded.type as string,
    func: decoded.func as string | undefined,
    args: decoded.args as unknown[] | undefined,
    stream: decoded.stream as boolean | undefined,
    result: decoded.result,
    error_type: decoded.error_type as string | undefined,
    message: decoded.message as string | undefined,
  };
}
